File: CSharp/DP.Domain/Samples/Proxy/ExtraPlacePricerProxy.cs
using System.Diagnostics;
using DP.Domain.Samples.Decorator;

namespace DP.Domain.Samples.Proxy
{
    public class ExtraPlacePricerProxy : IPricer
    {
        public string Customer { get; set; }
        public string Receiver { get; set; }
        public string Freight { get; set; }

        private readonly int MAX_EXTRA_PLACES = 2;
        private IPricer _extraPlacePricer = null;

        public ExtraPlacePricerProxy(IPricer pricer)
        {
            //The constructor's input parameter, pricer, is used for initialize a ExtraPlacePricer

            this._extraPlacePricer = new ExtraPlacePricer(pricer);
        }
        public decimal Price(Transport transport)
        {
            decimal totalPrice = 0;
            decimal servicePrice =0;
            var exceedMaxExtraPlaces = 0;
            
            if (transport.ExtraPlaceCnt <= MAX_EXTRA_PLACES)
            {
                totalPrice = this._extraPlacePricer.Price(transport);
            }
            else
            {
                exceedMaxExtraPlaces = transport.ExtraPlaceCnt - MAX_EXTRA_PLACES;//計算超過的加點
                transport.ExtraPlaceCnt = MAX_EXTRA_PLACES;
                totalPrice = this._extraPlacePricer.Price(transport);
                servicePrice = totalPrice * (decimal)0.2 * exceedMaxExtraPlaces; 
                totalPrice += servicePrice;
            }

            Trace.WriteLine($"加點(超過兩運送點)服務費用 = {servicePrice}，總費用={totalPrice}");
            return totalPrice;
        }
    }
}

File: Python/Samples/Proxy/ExtraPlacePricerProxy.py
import sys
import math
from decimal import *
import os.path
sys.path.append(os.path.join(os.path.dirname(__file__), '../Decorator/'))
from StdPricers import Pricer
from Models import Transport
from ServicePricers import ExtraPlacePricer


class ExtraPlacePricerProxy(Pricer):
    """加點服務計費
    """

    MAX_EXTRA_PLACES = 2

    def __init__(self, pricer=Pricer):
        super().__init__(pricer.customer, pricer.receiver, pricer.freight)
        self.extraPlacePricer = ExtraPlacePricer(pricer)

    def price(self, transport=Transport):
        """Return Total Price"""
        totalPrice = 0
        servicePrice = 0
        exceedMaxExtraPlaces = 0

        if (transport.extraPlaceCnt <= self.MAX_EXTRA_PLACES):
            totalPrice = self.extraPlacePricer.price(transport)
        else:
            exceedMaxExtraPlaces = transport.extraPlaceCnt - self.MAX_EXTRA_PLACES  # 計算超過的加點
            transport.extraPlaceCnt = self.MAX_EXTRA_PLACES
            totalPrice = self.extraPlacePricer.price(transport)
            servicePrice = Decimal(totalPrice * 0.2 * exceedMaxExtraPlaces)
            totalPrice += servicePrice

        print("加點(超過兩運送點)服務費用 = {0}，總費用={1}".format(servicePrice, totalPrice))
        return totalPrice
